File: app-core/src/main/java/com/topjohnwu/magisk/utils/SuConnector.java
package com.topjohnwu.magisk.utils;

import android.net.LocalSocket;
import android.net.LocalSocketAddress;
import android.os.Bundle;
import android.text.TextUtils;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

public abstract class SuConnector {

    private LocalSocket socket;
    protected DataOutputStream out;
    protected DataInputStream in;

    protected SuConnector(String name) throws IOException {
        socket = new LocalSocket();
        socket.connect(new LocalSocketAddress(name, LocalSocketAddress.Namespace.ABSTRACT));
        out = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream()));
        in = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
    }

    private String readString() throws IOException {
        int len = in.readInt();
        byte[] buf = new byte[len];
        in.readFully(buf);
        return new String(buf, "UTF-8");
    }

    public Bundle readSocketInput() throws IOException {
        Bundle bundle = new Bundle();
        while (true) {
            String name = readString();
            if (TextUtils.equals(name, "eof"))
                break;
            bundle.putString(name, readString());
        }
        return bundle;
    }

    public void response() {
        try {
            onResponse();
            out.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
        try {
            in.close();
            out.close();
            socket.close();
        } catch (IOException ignored) { }
    }

    protected abstract void onResponse() throws IOException;

}

File: app/src/main/java/com/topjohnwu/magisk/utils/SuConnector.kt
package com.topjohnwu.magisk.utils

import android.net.LocalSocket
import android.net.LocalSocketAddress
import android.os.Bundle
import android.text.TextUtils
import timber.log.Timber
import java.io.*
import java.nio.charset.Charset

abstract class SuConnector @Throws(IOException::class)
protected constructor(name: String) {

    private val socket: LocalSocket = LocalSocket()
    protected var out: DataOutputStream
    protected var input: DataInputStream

    init {
        socket.connect(LocalSocketAddress(name, LocalSocketAddress.Namespace.ABSTRACT))
        out = DataOutputStream(BufferedOutputStream(socket.outputStream))
        input = DataInputStream(BufferedInputStream(socket.inputStream))
    }

    @Throws(IOException::class)
    private fun readString(): String {
        val len = input.readInt()
        val buf = ByteArray(len)
        input.readFully(buf)
        return String(buf, Charset.forName("UTF-8"))
    }

    @Throws(IOException::class)
    fun readSocketInput(): Bundle {
        val bundle = Bundle()
        while (true) {
            val name = readString()
            if (TextUtils.equals(name, "eof"))
                break
            bundle.putString(name, readString())
        }
        return bundle
    }

    fun response() {
        runCatching {
            onResponse()
            out.flush()
        }.onFailure { Timber.e(it) }

        runCatching {
            input.close()
            out.close()
            socket.close()
        }
    }

    @Throws(IOException::class)
    protected abstract fun onResponse()

}
